### PR TITLE
revise T5 code to support one step decoding during generation

### PR DESCRIPTION
**Pull request by eelxpeng**
@thomwolf Hi, I'm new to contribute to this project. I revised the T5 code to support one step decoding during generation based on your implementation. Besides adding `decode_step` function, I also revised some others to pass in `cache` variable. Meanwhile, I added `bos_token` in tokenizer_t5 so that `bos_token` can be used as the first token during decoding. The example usage is as follows:
```
Examples:
            encoder_hidden_states = model.encode(input_ids)
            cache = model.init_state_from_encoder(encoder_hidden_states)
            next_token = input_ids.new_full((batch_size, 1), tokenizer.bos_token_id)
            generated = [next_token]
            for i in range(100):
                output, cache = model.decode_step(cache, input_ids=next_token)
                next_token = torch.argmax(logits, dim=-1).unsqueeze(-1)
                generated += [next_token]
            generated = torch.cat(generated, dim=1).tolist()
```
Let me know whether it is useful and can be merged.

### CODE REVIEW

File: src/transformers/tokenization_t5.py [hunk 1]
```diff
@@ -78,6 +78,7 @@ class T5Tokenizer(PreTrainedTokenizer):
     def __init__(
         self,
         vocab_file,
+        bos_token="<s>",
```
Comment (rpowalski) [inline src/transformers/tokenization_t5.py]: hi,
I think that this wouldn't work becouse `<s>` token does not exist in the original sentencepiece vocab (authors added only 3 special tokens - [link](https://github.com/google-research/text-to-text-transfer-transformer/blob/9fd7b14a769417be33bc6c850f9598764913c833/t5/data/sentencepiece_vocabulary.py#L35)). you could add `<s>` to the list of `additional_special_tokens` and then use `add_special_tokens` method to add it to encoder, but I think it is easier to use one of the existing extra_ids like shown below:
```suggestion
        bos_token="<extra_id_99>",
```

Comment (eelxpeng) [reply src/transformers/tokenization_t5.py]: Thank you for pointing out that. I agree with you that I shouldn't use '<s>' as the special token. In their original code, they actually use [ID 0](https://github.com/tensorflow/mesh/blob/b05b88c317ae78db6a678b3424fd24cc0982b1e2/mesh_tensorflow/transformer/transformer.py#L649) as the bos_token id. So I'm not sure if I should follow their way, or use another bos_token.

Comment (rpowalski) [reply src/transformers/tokenization_t5.py]: If you want to use padding for autoregressive decoder, I think, the best is to use `pad_token` instead of `bos_token`. In the original code 0 is defined as pad token (please have a look at the link in my previous comment)
